test/test_jujucontext.py

```diff
@@ -123,3 +126,214 @@ def test_parsing_all_str_fields(self):
         assert juju_context.secret_label == 'db-password'
         assert juju_context.unit_name == '42'
         assert juju_context.workload_name == 'workload'
+
+
+_valid_minimal = {
+    'JUJU_DISPATCH_PATH': 'hooks/install',
+    'JUJU_HOOK_NAME': 'install',
+    'JUJU_MODEL_NAME': 'foo',
+    'JUJU_MODEL_UUID': 'cdac5656-2423-4388-8f30-41854b4cca7d',
+    'JUJU_UNIT_NAME': '42',
+    'JUJU_VERSION': '3.4.0',
+}
+
+
+def test_context_from_os_environ(monkeypatch: pytest.MonkeyPatch):
+    with monkeypatch.context() as m:
+        m.setattr(os, 'environ', _valid_minimal)
+        juju_context = JujuContext.from_environ()
+    assert juju_context.dispatch_path == 'hooks/install'
+    assert juju_context.hook_name == 'install'
+    assert juju_context.model_name == 'foo'
+    assert juju_context.model_uuid == 'cdac5656-2423-4388-8f30-41854b4cca7d'
+    assert juju_context.unit_name == '42'
+    assert juju_context.version == JujuVersion('3.4.0')
+
+
+@pytest.mark.parametrize(
+    'event',
+    ['install', 'start', 'stop', 'remove', 'config-changed', 'update-status', 'leader-elected'],
+)
+@pytest.mark.parametrize(
+    'missing',
+    [
+        'JUJU_DISPATCH_PATH',
+        'JUJU_HOOK_NAME',
+        'JUJU_MODEL_NAME',
+        'JUJU_MODEL_UUID',
+        'JUJU_UNIT_NAME',
+        'JUJU_VERSION',
+    ],
+)
+def test_invalid_context_from_environ_simple(event: str, missing: str):
+    environ = _valid_minimal.copy()
+    environ['JUJU_HOOK_NAME'] = event
+    del environ[missing]
+    with pytest.raises(ValueError):
+        JujuContext.from_environ(environ)
+
+
+@pytest.mark.parametrize(
+    'event,additional_env',
+    [
+        ('secret-changed', {}),
+        ('secret-rotate', {}),
+        ('secret-remove', {'JUJU_SECRET_ID': 'secret:1'}),
+        ('secret-expired', {'JUJU_SECRET_ID': 'secret:1'}),
+        ('a-workload-pebble-ready', {}),
+        ('a-workload-pebble-check-failed', {'JUJU_WORKLOAD_NAME': 'a-workload'}),
+        ('a-workload-pebble-check-recovered', {'JUJU_PEBBLE_CHECK_NAME': 'chk1'}),
+        ('a-workload-pebble-custom-notice', {}),
+        ('storage-1-storage-attached', {}),
+        ('storage-1-storage-detaching', {}),
+        ('the-act', {'JUJU_DISPATCH_PATH': 'actions/the-act', 'JUJU_HOOK_NAME': ''}),
```

Comment:
I would have expected this way:

```suggestion
        ('', {'JUJU_DISPATCH_PATH': 'actions/the-act'}),
```

Is it the way it is so that the test name looks clean in pytest output only, or some other reason?

Reply:
Mostly, it's aligning with the other tests, where the event name is used more than once, and in the asserts. It does help with the pytest output as well though, as you mention - if I have only the dictionary here I think I would need to wrap these in Params so that it's more understandable which one failed if one does.